Add Vector Collection capacity planning guide [AI-327] (#1617)
hazelcast/hz-docs: docs/modules/data-structures/pages/vector-search-overview.adoc

To decrease pressure on heap memory, you can decrease the number of parallel migrations using `hazelcast.partition.max.parallel.migrations` and `hazelcast.partition.max.parallel.replications`.
====

== Capacity planning

=== Memory usage of Vector Collection

A few major elements contribute to Vector Collection memory requirements:

- entry keys and values: `memory footprint = entry count * (avg key size + avg value size)`.
The sizes should be calculated in serialized form.
- vectors: `memory footprint = entry count * dimensions * 4` (`4` being the size of the `float` coordinate).
The formula is for a collection with 1 vector index, with multiple vector indexes each would have to be added separately.

> Deduplication does not affect the memory usage significantly unless there are many duplicated vectors. In such case the footprint could be smaller.

Reviewer: ```suggestion
Deduplication does not affect the memory usage significantly unless there are many duplicated vectors. In such cases, the footprint could be smaller.
```

- vector index: `memory footprint = entry count * max degree * 8` (this is a simplification, but should be a good-enough approximation).
- additional overheads. They can be estimated as approximately `308` bytes per entry in collection with 1 vector index.
The precise value depends on architecture, JVM settings, if the deduplication is enabled and other factors.
The value given here is for Compressed OOPS and default 8-byte object alignment with deduplication enabled.

There are other factors and overheads, but they are smaller.

> When entries in Vector Collection are updated or removed, old vectors and nodes in vector index remain in memory until xref:vector-collections.adoc#optimize-collection[optimization] is performed. Enough heap space should be available for them.

Reviewer: ```suggestion
When entries in Vector Collection are updated or removed, old vectors and nodes in the vector index remain in memory until xref:vector-collections.adoc#optimize-collection[optimization] is performed. Enough heap space should be available for them.
```

> If backups are used (by default there is 1 backup), they need to be also included in the computation.

Reviewer: ```suggestion
If backups are used (by default there is 1 backup), they also need to be included in the computation.
```

Each backup (sync and async) needs the same amount of memory as primary data.
This brings us to the final formula for memory footprint:

`total memory footprint = (1 + backup count) * entry count * (avg key size + avg value size + dimensions * 4 + max degree * 8 + 308)`

In many cases, in particular with default settings and `dimensions > 1500` or with larger metadata (at least a 5kB), vector index size contributes less than 10% to the total memory usage.
In these cases the formula can be simplified and expressed in terms of raw data size (vectors and metadata):

`total memory footprint = (1 + backup count) * raw data size * 1.1`

All these data have to be distributed across the cluster members.
General recommendations from xref:ROOT:capacity-planning.adoc#memory-considerations[Hazelcast capacity planning] apply also to Vector Collection.
Keep enough headroom (40% of free heap memory is recommended) to accommodate member failure or shutdown and ensure that the cluster has enough resources to xref:ROOT:capacity-planning.adoc#sizing-for-failures[operate after member failures].

> If all members have the same heap size, the vector collection will need cluster at least of size:

Reviewer: ```suggestion
If all members have the same heap size, the vector collection will need a cluster at least of size:
```

`number of members = ceil(total memory footprint / (heap size * 60%))`

> If other data will be also stored in the cluster, eg. IMaps, they also have to be accounted for.

Reviewer: ```suggestion
If other data will also be stored in the cluster, eg. IMaps, this also has to be accounted for.
```
